src/IoTeX/Address.cpp

```diff
@@ -0,0 +1,65 @@
+// Copyright © 2017-2019 Trust Wallet.
+//
+// This file is part of Trust. The full Trust copyright notice, including
+// terms governing use, modification, and redistribution, is contained in the
+// file LICENSE at the root of the source code distribution tree.
+
+#include "Bech32.h"
+#include "Hash.h"
+
+#include "Address.h"
+
+using namespace TW;
+using namespace TW::IoTeX;
+
+bool Address::isValid(const std::string& addr) {
+    if (addr.size() != 41 || addr[0] != 'i' || addr[1] != 'o') {
+        return false;
+    }
+
+    const auto dec = Bech32::decode(addr);
+    if (dec.second.empty()) {
+        return false;
+    }
+    if (dec.first != "io") {
+        return false;
+    }
+
+    Data keyHash;
+    auto success = Bech32::convertBits<5, 8, false>(keyHash, dec.second);
+    if (!success || keyHash.size() != 20) {
+        return false;
+    }
+    return true;
+}
+
+Address::Address(const std::string& addr) {
+    if (!isValid(addr)) {
+        throw std::invalid_argument("IoTeX: Invalid address data");
+    }
+
+    const auto dec = Bech32::decode(addr);
+    Bech32::convertBits<5, 8, false>(keyHash, dec.second);
+}
+
+Address::Address(const PublicKey& publicKey) {
+    if (publicKey.type != TWPublicKeyTypeSECP256k1Extended) {
+        throw std::invalid_argument("IoTeX: Address needs an extended SECP256k1 public key");
+    }
+    keyHash = publicKey.hash({}, static_cast<Data(*)(const byte*, const byte*)>(Hash::keccak256), true);
+    keyHash.erase(keyHash.begin(), keyHash.begin() + 12);
```

Comment:
public key hash is the last 20-byte of actual hash, similar to Ethereum. hence remove first 12 bytes

Reply:
is this bech32 address? then we have an address for it, `Bech32Address` or `TendermintAddress`

Reply:
yes, the address is Bech32 encoding of the public key hash, you can please refer to Address::string()

Reply:
will update the PR soon, thx

Reply:
Just updated, it's called `CosmosAddress`, and there is also `SegwitAddress` which is bech32 for bitcoin

Reply:
yes you are right, our address is similar, with different hrp
the implementation referenced these existing code